Sources/Sentry/SentryUIViewControllerPerformanceTracker.m

```diff
@@ -163,25 +163,51 @@ - (void)startRootSpanFor:(UIViewController *)controller
         return;
     }
 
-    [self.currentTTDTracker finishSpansIfNotFinished];
-
     SentryTimeToDisplayTracker *ttdTracker =
-        [[SentryTimeToDisplayTracker alloc] initForController:controller
-                                           waitForFullDisplay:self.enableWaitForFullDisplay
-                                         dispatchQueueWrapper:_dispatchQueueWrapper];
+        [self startTimeToDisplayTrackerForScreen:[SwiftDescriptor getObjectClassName:controller]
+                              waitForFullDisplay:self.alwaysWaitForFullDisplay
+                                          tracer:(SentryTracer *)vcSpan];
 
-    if ([ttdTracker startForTracer:(SentryTracer *)vcSpan]) {
+    if (ttdTracker) {
         objc_setAssociatedObject(controller, &SENTRY_UI_PERFORMANCE_TRACKER_TTD_TRACKER, ttdTracker,
             OBJC_ASSOCIATION_ASSIGN);
-        self.currentTTDTracker = ttdTracker;
-    } else {
-        self.currentTTDTracker = nil;
     }
 }
 
 - (void)reportFullyDisplayed
 {
-    [self.currentTTDTracker reportFullyDisplayed];
+    SentryTimeToDisplayTracker *tracker = self.currentTTDTracker;
+    if (tracker) {
+        if (tracker.waitForFullDisplay) {
+            [self.currentTTDTracker reportFullyDisplayed];
+        } else {
+            SENTRY_LOG_WARN(@"Transaction is not waiting for full display report. You can enable "
+                            @"`enableTimeToFullDisplay` option, or use the waitForFullDisplay "
+                            @"property in our `SentryTracedView` view for SwiftUI.");
+        }
+    } else {
+        SENTRY_LOG_DEBUG(@"No screen transaction being tracked right now.")
+    }
+}
```

Comment:
`l:` this method could be rewritten using Early Return pattern, to reduce nesting

```suggestion
    SentryTimeToDisplayTracker *tracker = self.currentTTDTracker;
    if (tracker == nil) {
        SENTRY_LOG_DEBUG(@"No screen transaction being tracked right now.")
        return;
    }
    if (!tracker.waitForFullDisplay) {
        SENTRY_LOG_WARN(@"Transaction is not waiting for full display report. You can enable "
                        @"`enableTimeToFullDisplay` option, or use the waitForFullDisplay "
                        @"property in our `SentryTracedView` view for SwiftUI.");
        return;
    }
    [self.currentTTDTracker reportFullyDisplayed];
```